src/core/server/ui_settings/ui_settings_client.ts

```diff
@@ -147,17 +143,50 @@ export class UiSettingsClient implements IUiSettingsClient {
     return defaultsDeep(userProvided, this.defaults);
   }
 
+  private validateKey(key: string, value: unknown) {
+    const definition = this.defaults[key];
+    if (value === null || definition === undefined) return;
+    if (definition.schema) {
+      definition.schema.validate(value, {}, `validation [${key}]`);
+    }
+  }
+
+  private onWriteHook(changes: Record<string, unknown>) {
+    for (const key of Object.keys(changes)) {
+      this.assertUpdateAllowed(key);
+    }
+
+    for (const [key, value] of Object.entries(changes)) {
+      this.validateKey(key, value);
+    }
+  }
+
+  private onReadHook<T = unknown>(values: Record<string, unknown>) {
+    // write the userValue for each key stored in the saved object that is not overridden
+    // validate value read from saved objects as it can be changed via SO API
+    const filteredValues: UserProvided<T> = {};
+    for (const [key, userValue] of Object.entries(values)) {
+      if (userValue === null || this.isOverridden(key)) continue;
+      try {
+        this.validateKey(key, userValue);
+        filteredValues[key] = {
+          userValue: userValue as T,
+        };
+      } catch (error) {
```

Comment:
aside: the more places I see use config-schema, the more I think it shouldn't throw exception messages. Invalid data isn't really an exception, if anything, it's almost entirely expected.

Reply:
Yeah. In my initial implementation, I had to wrap it in a function returning `{error?: string, valid: boolean}`. I'd prefer @kbn/config-schema to have such an interface.

Reply:
@joshdover should I create an issue to investigate refactoring options?